ipam: Fix invalid PodCIDR in CiliumNode in ENI/Azure/MultiPool mode

gandro opened the pull request:
This commit fixes an issue where Cilium-Agent would announce its internal auto-generated IPv4/IPv6AllocRange by writing it into the local CiliumNode resource.

Cilium-Agents IPv4/IPv6AllocRange is Cilium's internal representation of the local node's PodCIDR. The concept of a single allocation CIDR however is outdated, since in many IPAM modes (ENI, Azure, AlibabaCloud, MultiPool) there either is no PodCIDR to begin with (and pod IPs are allocated using a different scheme), or there are multiple PodCIDRs (in the case of MultiPool).

Before this commit, cilium-agent used to copy its internal PodCIDR into the CiliumNode resource, but this behavior is only correct in Kubernetes IPAM mode. In Kubernetes IPAM mode, the IPv4/IPv6AllocRange is taken from the Kubernetes Node resource, and thus it makes sense to announce it to other cilium-agent instances via the CiliumNode resource. This field then is for example used by `auto-direct-node-routes` to install routes for direct routing mode.

However, in all other IPAM modes that we have, copying the internal PodCIDR into the the CiliumNode resource does not make sense:

 - In ClusterPool mode, the CiliumNode PodCIDR field is populated by cilium-operator. Therefore, we should try to overwrite it ourselves. This behavior is unchanged by this commit.
 - In all other current IPAM modes, the CiliumNode PodCIDR field is not used. Instead we use fields specific to those IPAM modes (c.f. CiliumNode.Spec.IPAM). In those modes, the previous code therefore wrote a auto-generated but otherwise IPv4/IPv6AllocRange into the field. That auto-generated PodCIDR is not used for pod IP allocation in those modes, so it does not make sense to announce it either.

The announced auto-generated PodCIDR is causing issues in IPAM MultiPool mode. With this commit, we will no longer install invalid routes for the auto-generated PodCIDR. This was mostly harmless (since the auto-generated PodCIDR rarely overlapped with the real ones), but still a potential cause for confusion and bugs.

In addition, ENI-mode users have reported that they are confused by the this "fake" PodCIDR, which is understandable, since that PodCIDR is meaningless in ENI mode (#9409).

Long term, we want to remove the concept of the IPv4/IPv6AllocCIDR completely (and remove the auto-generation code), but this is a first incremental step towards that goal.

Fixes: #9409

```release-note
Fix potential cross-node connectivity issue when IPsec is enabled with ENI or Azure IPAM modes.
```

gandro committed:
ipam: Fix invalid PodCIDR in CiliumNode in ENI/Azure/MultiPool mode

This commit fixes an issue where Cilium-Agent would announce its
internal auto-generated IPv4/IPv6AllocRange by writing it into the local
CiliumNode resource.

Cilium-Agents IPv4/IPv6AllocRange is Cilium's internal representation of
the local node's PodCIDR. The concept of a single allocation CIDR
however is outdated, since in many IPAM modes (ENI, Azure, AlibabaCloud,
MultiPool) there either is no PodCIDR to begin with (and pod IPs are
allocated using a different scheme), or there are multiple PodCIDRs (in
the case of MultiPool).

Before this commit, cilium-agent used to copy its internal PodCIDR into
the CiliumNode resource, but this behavior is only correct in Kubernetes
IPAM mode. In Kubernetes IPAM mode, the IPv4/IPv6AllocRange is taken
from the Kubernetes Node resource, and thus it makes sense to announce
it to other cilium-agent instances via the CiliumNode resource. This
field then is for example used by `auto-direct-node-routes` to install
routes for direct routing mode.

However, in all other IPAM modes that we have, copying the internal
PodCIDR into the CiliumNode resource does not make sense:

 - In ClusterPool mode, the CiliumNode PodCIDR field is populated by
   cilium-operator. Therefore, we should not try to overwrite it
   ourselves.  This behavior is unchanged by this commit.
 - In all other current IPAM modes, the CiliumNode PodCIDR field is not
   used. Instead we use fields specific to those IPAM modes (c.f.
   CiliumNode.Spec.IPAM). In those modes, the previous code therefore wrote
   a auto-generatead, but otherwise valid IPv4/IPv6AllocRange into the
   field. That auto-generated PodCIDR is not used for pod IP allocation in
   those modes, so it does not make sense to announce it either.

The announced auto-generated PodCIDR is causing issues in IPAM MultiPool
mode. With this commit, we will no longer install invalid routes for the
auto-generated PodCIDR. This was mostly harmless (since the
auto-generated PodCIDR rarely overlapped with the real ones), but still
a potential cause for confusion and bugs.

In addition, ENI-mode users have reported that they are confused by the
this "fake" PodCIDR, which is understandable, since that PodCIDR is
meaningless in ENI mode (#9409).

Long term, we want to remove the concept of the IPv4/IPv6AllocCIDR
completely (and remove the auto-generation code), but this is a first
incremental step towards that goal.

Fixes: #9409

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

thorn3r reviewed (approved):
thanks for the detailed commit msg, looks great :+1:

gandro committed:
ipsec: Do not use AllocCIDR with subnet encryption

IPSec relies on the "pod subnet" CIDRs for encryption in IPAM modes
(e.g. ENI, Azure) which do not have a pod CIDR (aka. alloc CIDR).

Therefore, we should not use the alloc CIDR in such modes. This commit
moves the setup of routes based on the alloc CIDR into the conditions
which are executed if we do are not in pod subnet based encryption.

In addition, we add some temporary code to remove the old bogus route
based for the local node in subnet encryption mode.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

pchaigno commented on pkg/nodediscovery/nodediscovery.go:
```diff
@@ -507,17 +507,16 @@ func (n *NodeDiscovery) mutateNodeResource(nodeResource *ciliumv2.CiliumNode) er
 		}
```
*(On commit description)*

(Not sure if it's worth repushing just for that.)

> In ClusterPool mode, the CiliumNode PodCIDR field is populated by cilium-operator. Therefore, we should try to overwrite it ourselves.

Did you mean "we shouldn't"?

> the previous code therefore wrote a auto-generated but otherwise IPv4/IPv6AllocRange into the field

I think there's a word missing here.

gandro replied:
> Did you mean "we shouldn't"?

> I think there's a word missing here.

Yes, good catches! Will fix, thanks!

pchaigno commented on pkg/datapath/linux/node.go:
```diff
@@ -1081,26 +1081,29 @@ func (n *linuxNodeHandler) enableIPsecIPv4(newNode *nodeTypes.Node, nodeID uint1
 	errs = errors.Join(errs, upsertIPsecLog(err, "default-drop IPv4", wildcardCIDR, wildcardCIDR, spi))
 
 	if newNode.IsLocal() {
-		errs = errors.Join(errs, n.replaceNodeIPSecInRoute(new4Net))
```
IIUC, on EKS & AKS, we would install a route using the internal auto-generated IPv4/IPv6AllocRange. Do we have an idea of the potential consequences of that? The second commit suggests that the auto-generated CIDR could clash with real CIDR, although unlikely.

gandro replied:
So yes, before this PR installed a route for the auto-generated pod CIDR. It used an auto-generated pod CIDR, because that's Cilium's ancient fall-back if no pod CIDR is configured. That auto-generated pod CIDR is not used anywhere (see also https://github.com/cilium/cilium/issues/27018) and we're planning to eventually remove it, since it causes confusion for both engineers and users.

Particularly in EKS and AKS, IPs are not allocated from any pod CIDR (be it auto-generated or manually configured), and so using a node's pod CIDR in AKS or EKS is making assumptions about the IP range that does not hold true. By installing a route for each node's (auto-generated) pod CIDR (as we did before the PR), we routed all traffic belonging to those auto-generated CIDRs to the encrypt device.  

I'm not sure, what the consequence of this were. We have not received bug reports about this, I think mostly because the auto-generated CIDR is by design unlikely to clash with a real CIDR. Since the encrypt device is probably also the default egress device, the route might also not have affected any issues for CIDRs that were routable on the default egress device anyway.

In any case, my primary motivation for this PR is the clean up unconditional uses of "pod CIDR" in the agent code, because such code will not be compatible with pod-CIDR-less IPAM modes such as EKS/AKS or Multi-Pool. IPSec already has "subnet encryption" as a pod CIDR replacement, but for some reason it still unconditionally used the pod CIDR, which this PR attempts to fix.

pchaigno replied:
I did a quick test on EKS to simulate the impact of a clash between the auto-generated CIDR and the pod subnet. Pod subnet was 192.168.0.0/16 and I manually installed a route like the one removed in this PR but with CIDR 192.168.128.0/17.

That caused some existing connections to be interrupted and prevented me from running the connectivity tests. I didn't debug further. There were no drops reported by Cilium or XFRM errors.

I have no idea how unlikely it is for the CIDRs to clash, but maybe we need to add a release note (something like below or can we make the "potential" more precise?) and backport to just v1.14?
```
Fix potential cross-node connectivity issue when IPsec is enabled with ENI or Azure IPAM modes.
```

gandro replied:
Awesome, thanks for that test :rocket: 

> I have no idea how unlikely it is for the CIDRs to clash, but maybe we need to add a release note (something like below or can we make the "potential" more precise?) and backport to just v1.14?

In the end, it depends on user configuration. [The auto-generated CIDR](https://github.com/cilium/cilium/blob/a8d80e25ca2d26affe4f7f7a8350bb537ddbe388/pkg/node/address.go#L123) is of the form "10.x.0.0/16" where x is derived from the node's IP. So I think clashes can happen - but I'm not aware of any bug reports, probably because most users are using IPs in the range of "10.0.0.0/16".

I think we can absolutely backport this to v1.14. I removed the backport label because there have not been bug reports and this has likely been present for years, but given the risk should be low, it does fit our backporting policy.

pchaigno reviewed (approved):
Thanks a lot for delaying this until we can have enough confidence into the IPsec CI and changes :pray: I know it's a pain to have to postpone PRs like that.

I have one comment below to try to understand the potential impact of the IPsec change. But regardless of that, I see the PR is marked as release-note/bug, but doesn't have a release note or a backport label. Outside of the IPsec part, is there any user-visible impact?
